text/0000-macro-metavar-expr.md

```diff
@@ -0,0 +1,332 @@
+
+## Count
+
+The `${count(x)}` metavariable expression shown in the `vec` example in the
+previous section counts the number of repetitions that will occur if the
+identifier is used in a repetition at this depth.  This means that in a macro
+expansion like:
+
+```
+    ${count(x)} $( $x )*
+```
+
+the expression `${count(x)}` will expand to the number of times the `$( $x )*`
+repetition will repeat.
+
+If repetitions are nested, then an optional depth parameter can be used to
+limit the number of nested repetitions that are counted.  For example, a macro
+expansion like:
+
+```
+    ${count(x, 1)} ${count(x, 2)} ${count(x, 3)} $( a $( b $( $x )* )* )*
+```
+
+The three values this expands to are the number of outer-most repetitions (the
+number of times `a` would be generated), the sum of the number of middle
+repetitions (the number of times `b` would be generated), and the total number
+of repetitions of `$x`.
+
+## Index and length
+
+Within a repetition, the `${index()}` and `${length()}` metavariable
+expressions give the index of the current repetition and the length of the
+repetition (i.e., the number of times it will repeat). The index value ranges
+from `0` to `length - 1`.
+
+For nested repetitions, the `${index()}` and `${length()}` metavariable
+expressions expand to the inner-most index and length respectively.
+If the `depth` parameter is specified, then the metavariable expression
+expands to the index or length of the surrounding nested repetition, counting
+outwards from the inner-most repetition.  The expressions `${index()}` and
+`${index(0)}` are equivalent.
+
+For example in the expression:
+
+```
+    $( a $( b $( c $x ${index()}/${length()} ${index(1)}/${length(1)} ${index(2)}/${length(2)} )* )* )*
+```
+
+the first pair of values are the index and length of the inner-most
+repetition, the second pair are the index and length of the middle
+repetition, and the third pair are the index and length of the outer-most
+repetition.
+
+## Ignore
+
+Sometimes it is desired to repeat an expansion the same number of times as a
+metavariable repeats but without actually expanding the metavariable.  It may
+be possible to work around this by expanding the metavariable in an expression
+like `{ $x ; 1 }`, where the expanded value of `$x` is ignored, but this
+is only possible if what `$x` expands to is valid in this kind of expression.
+
+The `${ignore(ident)}` metavariable acts as if `ident` was used for the purposes
+of repetition, but expands to nothing.  This means a macro expansion like:
+
+```
+    $( ${ignore(x)} a )*
+```
+
+will expand to a sequence of `a` tokens repeated the number of times that `x` repeats.
+
+## Dollar dollar
+
+Since metavariable expressions always apply during the expansion of the macro,
+they cannot be used in recursive macro definitions.  To allow recursive macro
+definitions to use metavariable expressions, the `$$` expression expands to a
+single `$` token.
```

Comment:
Are they any places where this would want to go deeper?  Would it be helpful to have `$$$` that expands to `$$`, instead of needing `$$$$`?  (Does `$$$$` work with this RFC, actually?  Is exponential escaping bad?)

Reply:
Going deeper is only necessary if macro definitions are multiple-times recursive (a macro that defines a macro that defines a macro), and you want to defer metavariable expansions or repetitions to the inner macros in ways that are otherwise ambiguous.  The doubling up of the escape characters for each level is necessary so that at each nesting level you can represent a meta-variable whose name is stored in another meta-variable. An even number of `$` followed by `var` (e.g. `$$$$var`) expands to `n/2` `$`s followed by a literal var (e.g. `$$var`). An odd number of `$` expands to `(n-1)/2` `$`s followed by the expansion of `$var`, (e.g. if `$var` == `foo` then `$$$$$var` expands to `$$foo`).

An example of where this would be necessary in existing code is [here](https://github.com/bluss/defmac/blob/6886f04d412e1ee2b6d4d240feb51e51a9caf808/src/lib.rs#L87). This code is currently using `$dol` as a hack for what `$$` would provide, and `$dol $arg` would become `$$$arg`.

This is the same as for `\`-escaping in strings, and most other kinds of escaping in other languages, so it should be familiar to users.

Although sixteen `$` in a row wouldn't be great, quadruply-recursive macro definitions are probably not a great idea either, and it should be possible to break the macro down into separate parts with less nesting if that does become a concern.